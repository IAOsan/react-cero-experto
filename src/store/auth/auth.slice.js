import { createSlice } from '@reduxjs/toolkit';
import userService from '../../services/user.service';
import authService from '../../services/auth.service';
import {
	loadingStatus,
	successStatus,
	errorStatus,
	setError,
} from '../ui/ui.slice';

const slice = createSlice({
	name: '[auth]',
	initialState: {
		checking: true,
		isAuth: false,
		user: null,
	},
	reducers: {
		logged_in(state, { payload }) {
			state.checking = false;
			state.isAuth = true;
			state.user = payload;
		},
		logged_out(state) {
			state.isAuth = false;
			state.user = null;
		},
		checking_auth_started(state) {
			state.checking = true;
		},
		checking_auth_finished(state) {
			state.checking = false;
		},
	},
});

const { logged_in, logged_out, checking_auth_started, checking_auth_finished } =
	slice.actions;
export default slice.reducer;

/* -------------------------------------------------------------------------- */
/*                               action creators                              */
/* -------------------------------------------------------------------------- */
export const registerEmailAndPassword = (user) => async (dispatch) => {
	dispatch(loadingStatus());
	try {
		const res = await userService.register(user);
		dispatch(successStatus());
		dispatch(logged_in(res));
	} catch (error) {
		dispatch(errorStatus());
		dispatch(setError({ register: error }));
	}
};

export const loginEmailAndPassword = (user) => async (dispatch) => {
	dispatch(loadingStatus());
	try {
		const res = await authService.login(user);
		dispatch(successStatus());
		dispatch(logged_in(res));
	} catch (error) {
		dispatch(errorStatus());
		dispatch(setError({ login: error }));
	}
};

export const logout = () => (dispatch) => {
	try {
		authService.logout();
		dispatch(logged_out());
	} catch (error) {
		dispatch(setError(error));
	}
};

export const checkAuth = () => async (dispatch) => {
	const user = await authService.getCurrentUser();

	dispatch(checking_auth_started());
	if (user) {
		dispatch(logged_in(user));
	}
	dispatch(checking_auth_finished());
};

/* -------------------------------------------------------------------------- */
/*                                  selectors                                 */
/* -------------------------------------------------------------------------- */
export const selectAuthState = (state) => state.auth;
