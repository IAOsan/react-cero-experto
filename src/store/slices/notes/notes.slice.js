import { createSlice } from '@reduxjs/toolkit';
import { toast } from 'react-toastify';
import * as notesService from '../../../services/notes.service';
import * as imagesService from '../../../services/images.service';
import { convertFileToBase64 } from '../../../constants/utils';

const initialState = {
	list: [],
	active: null,
};

const slice = createSlice({
	name: '[notes]',
	initialState,
	reducers: {
		note_added(state, { payload }) {
			state.list.push(payload);
		},
		note_updated(state, { payload }) {
			const { id } = state.active;
			state.list = state.list.map((n) =>
				n.id === id ? { ...n, ...payload } : n
			);
		},
		note_changed(state, { payload }) {
			state.active = {
				...state.active,
				...payload,
			};
		},
		note_removed(state, { payload }) {
			state.list = state.list.filter((n) => n.id !== payload);
			state.active = null;
		},
		note_selected(state, { payload }) {
			const activeNote = state.list.find((n) => n.id === payload);
			state.active = activeNote;
		},
		notes_loaded(state, { payload }) {
			state.list = payload;
		},
		notes_cleared(state) {
			Object.keys(initialState).forEach((k) => {
				state[k] = initialState[k];
			});
		},
	},
});

export default slice.reducer;

const {
	note_added,
	note_changed,
	note_updated,
	note_selected,
	notes_loaded,
	note_removed,
	notes_cleared,
} = slice.actions;

export const addNote = async (dispatch, getState) => {
	try {
		const { uid } = getState().auth.user;
		const note = await notesService.addNewNote(uid);
		dispatch(note_added(note));
		dispatch(note_selected(note.id));
	} catch (error) {
		toast.error(error);
	}
};

export const changeNote = note_changed;

export const updateNote = (note) => async (dispatch, getState) => {
	try {
		const { uid } = getState().auth.user;

		if (note.file) {
			const meta = await imagesService.uploadImage(note.file);
			Object.assign(note, { imageUrl: meta.secure_url });
			delete note.file;
		}

		await notesService.updateNote(uid, note);
		dispatch(note_updated(note));
		toast.success('👏 Note saved!');
	} catch (error) {
		toast.error(error);
	}
};

export const selectNote = note_selected;

export const loadNotes = (uid) => async (dispatch) => {
	try {
		const notes = await notesService.getNotes(uid);
		dispatch(notes_loaded(notes));
	} catch (error) {
		toast.error(error);
	}
};

export const addImageToNote = (file) => async (dispatch) => {
	const base64 = await toast.promise(convertFileToBase64(file), {
		pending: 'Adding image, please wait...',
		success: 'Image added 👌',
		error: 'Sorry, the image could not be added 🤯, try uploading another one or try again later',
	});
	dispatch(note_changed({ imageUrl: base64 }));
};

export const removeNote = (id) => async (dispatch, getState) => {
	try {
		const { uid } = getState().auth.user;
		await notesService.removeNote(uid, id);
		dispatch(note_removed(id));
		toast.success('👍 Note removed!');
	} catch (error) {
		toast.error(error);
	}
};

export const clearNotes = notes_cleared();
