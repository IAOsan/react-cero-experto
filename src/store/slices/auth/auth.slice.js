import { createSlice } from '@reduxjs/toolkit';
import {
	startLoading,
	finishLoading,
	addError,
	removeError,
} from '../ui/ui.slice';
import { clearNotes } from '../notes/notes.slice';
import * as authService from '../../../services/auth.service';

const initialState = {
	user: null,
	isAuth: false,
};

const slice = createSlice({
	name: '[auth]',
	initialState,
	reducers: {
		logged_in(state, { payload }) {
			state.user = payload;
			state.isAuth = true;
		},
		logged_out(state, action) {
			state.user = null;
			state.isAuth = false;
		},
	},
});

export default slice.reducer;

const { logged_in, logged_out } = slice.actions;

export const login = logged_in;

export const loginWithEmailPassWord = (data) => async (dispatch) => {
	dispatch(startLoading);
	dispatch(removeError);
	try {
		const user = await authService.loginEmailPassword(data);
		dispatch(logged_in(user));
	} catch (error) {
		dispatch(addError(error));
	} finally {
		dispatch(finishLoading);
	}
};

export const loginWithGoogle = async (dispatch) => {
	dispatch(removeError);
	try {
		const user = await authService.loginGoogle();
		dispatch(logged_in(user));
	} catch (error) {
		dispatch(addError(error));
	}
};

export const registerWithEmailPassword = (data) => async (dispatch) => {
	dispatch(startLoading);
	dispatch(removeError);
	try {
		const user = await authService.registerEmailPassword(data);
		dispatch(logged_in(user));
	} catch (error) {
		dispatch(addError(error));
	} finally {
		dispatch(finishLoading);
	}
};

export const logout = async (dispatch) => {
	await authService.logout();
	dispatch(logged_out());
	dispatch(clearNotes);
};
