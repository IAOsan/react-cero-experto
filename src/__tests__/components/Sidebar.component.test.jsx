import '@testing-library/jest-dom';
import { vi } from 'vitest';
import { renderWithProviders, setupUser } from '../test-utils';
import Sidebar from '../../components/Sidebar.component';
import { addNote } from '../../store/slices/notes/notes.slice';
import { logout } from '../../store/slices/auth/auth.slice';

vi.mock('../../store/slices/auth/auth.slice', async () => {
	const org = await vi.importActual('../../store/slices/auth/auth.slice');
	return {
		...org,
		logout: vi.fn(),
	};
});

vi.mock('../../store/slices/notes/notes.slice', async () => {
	const org = await vi.importActual('../../store/slices/notes/notes.slice');
	return {
		...org,
		addNote: vi.fn(),
	};
});

const userName = 'test user';
const user = setupUser();

describe('<Sidebar />', () => {
	beforeEach(() => {
		vi.clearAllMocks();
	});

	it('should renders properly', () => {
		const wrapper = renderSidebar();
		expect(wrapper.container.firstElementChild).toMatchSnapshot();
		expect(wrapper.getByText(userName)).toMatchSnapshot();
		expect(
			wrapper.getByRole('button', { name: /logout/i })
		).toBeInTheDocument();
		expect(
			wrapper.getByRole('button', { name: /new entry/i })
		).toBeInTheDocument();
		expect(wrapper.getByRole('list')).toBeInTheDocument();
	});

	it('should be able to add note when clicks new entry btn', async () => {
		const wrapper = renderSidebar();

		await user.click(wrapper.getByRole('button', { name: /new entry/i }));

		expect(addNote).toHaveBeenCalled();
	});

	it('should logout when clicks logout btn', async () => {
		const wrapper = renderSidebar();

		await user.click(wrapper.getByRole('button', { name: /logout/i }));

		expect(logout).toHaveBeenCalled();
	});
});

function renderSidebar() {
	return renderWithProviders(<Sidebar />, {
		preloadedState: {
			auth: {
				user: {
					name: userName,
				},
				isAuth: true,
			},
		},
	});
}
