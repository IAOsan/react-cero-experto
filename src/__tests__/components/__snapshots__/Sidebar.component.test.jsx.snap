// Vitest Snapshot v1

exports[`<Sidebar /> > should renders properly 1`] = `
<aside
  class="bg-dark home__sidebar sidebar anima-slide-in-left"
>
  <div
    class="flex flex-ai-c flex-jc-sb"
  >
    <p>
      <span>
        <svg
          class="icon mr-8"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      test user
    </p>
    <button
      class="btn btn--default btn--dark"
      type="button"
    >
      Logout
    </button>
  </div>
  <button
    class="btn my-12 text-center sidebar__new-entry"
    type="button"
  >
    <svg
      class="icon icon-display-3"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    <p>
      New entry
    </p>
  </button>
  <ul
    class="spacing-y-8 entries"
  />
</aside>
`;

exports[`<Sidebar /> > should renders properly 2`] = `
<p>
  <span>
    <svg
      class="icon mr-8"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </span>
  test user
</p>
`;
